### PR TITLE
Edge: Support context menu handling via ICoreWebView2_11

### PR DESCRIPTION
**Pull request by sratz**
Support custom context menus and suppressing of context menus by implementing ICoreWebView2_11#add_ContextMenuRequested().

### CODE REVIEW

**Comment by sratz** [comment]
@HeikoKlare @amartya4256
This change is ready for review. Would be great if you could have a look!
The implementation is based on IE (`WebSite.ShowContextMenu(int, long, long, long)`)

File: bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java [hunk 1]
```diff
@@ -698,6 +706,35 @@ int handleDOMContentLoaded(long pView, long pArgs) {
...
+	args.get_Location(locationPointer);
+	POINT pt = new POINT();
+	OS.MoveMemory(pt, locationPointer, POINT.sizeof);
+	pt.x = DPIUtil.autoScaleDown(pt.x); // To Points
+	pt.y = DPIUtil.autoScaleDown(pt.y); // To Points
+	Event event = new Event();
+	event.x = pt.x;
+	event.y = pt.y;
+	browser.notifyListeners(SWT.MenuDetect, event);
+	if (event.doit) {
```
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java:722]: Currently, it behaves the opposite of how it is supposed to behave. I think this needs to be inverted (to also align with the IE implementation in `WebSite`):
```suggestion
	if (!event.doit) {
```

Comment (sratz) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java:722]: oops

File: bundles/org.eclipse.swt/Eclipse SWT PI/win32/org/eclipse/swt/internal/ole/win32/ICoreWebView2_11.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     SAP SE - initial implementation
+ *******************************************************************************/
+package org.eclipse.swt.internal.ole.win32;
+
+public class ICoreWebView2_11 extends ICoreWebView2 {
```
[the remark this answers is not in the document]
Comment (sratz) [reply bundles/org.eclipse.swt/Eclipse SWT PI/win32/org/eclipse/swt/internal/ole/win32/ICoreWebView2_11.java:16]: Oh, yes, good idea.

**Review by HeikoKlare — APPROVED** [review]
Looks good now. Ready to merge from my side.

File: bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java [hunk 3]
```diff
@@ -698,6 +706,35 @@ int handleDOMContentLoaded(long pView, long pArgs) {
...
 
+int handleContextMenuRequested(long pView, long pArgs) {
+	ICoreWebView2ContextMenuRequestedEventArgs args = new ICoreWebView2ContextMenuRequestedEventArgs(pArgs);
+
+	long[] locationPointer = new long[1];
+	args.get_Location(locationPointer);
+	POINT pt = new POINT();
+	OS.MoveMemory(pt, locationPointer, POINT.sizeof);
+	pt.x = DPIUtil.autoScaleDown(pt.x); // To Points
+	pt.y = DPIUtil.autoScaleDown(pt.y); // To Points
```
[the remark this answers is not in the document]
Comment (akoch-yatta) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java:717]: @HeikoKlare Shouldn't the nativeZoom of the browser shell work here as @amartya4256 suggested?

Other question regarding this method args.get_Location will return "the coordinates where the context menu request occurred in relation to the upper left corner of the WebView bounds.". Shouldn't that be mapped to the display coordinates before passing it to event.x and event.y? menu.setLocation will only work with display (point) coordinates.

[the remark this answers is not in the document]
Comment (sratz) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java:717]: From `Event#x`s Javadoc:

_the **widget-relative**, x coordinate of the pointer at the time the mouse button was pressed or released_

So widget-relative coordinates in the `Event` are correct.

`Menu#setLocation(int, int)` is not relevant here, it is never called (`Menu#x/y` are both 0 and `Menu#hasLocation` is `false`).

If someone were to call `Menu#setLocation(int, int)`, then they'd have to do the coordinate transformation.

[the remark this answers is not in the document]
Comment (akoch-yatta) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java:717]: Ah, yes, you are right. I found a usage, where it was used as absolute value, so I didn't look into the javadoc but you are right.
Still, then we need to do the mapping to display coordinates before using setLocation in line 730. don't we?

[the remark this answers is not in the document]
Comment (HeikoKlare) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java:717]: > Shouldn't the nativeZoom of the browser shell work here as @amartya4256 suggested?

Sounds reasonable to me. I was not aware that a `WebBrowser` knows about the using `Browser`, but in that case it makes sense to just retrieve the information from the `Browser` widget.

[the remark this answers is not in the document]
Comment (sratz) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java:717]: Hmm, good point.

I guess the same would apply to IE / `org.eclipse.swt.browser.WebSite.ShowContextMenu(int, long, long, long)` as well.

[the remark this answers is not in the document]
Comment (sratz) [reply bundles/org.eclipse.swt/Eclipse SWT Browser/win32/org/eclipse/swt/browser/Edge.java:717]: Ok, this is strange...
TypedListener converts a regular Event to a MenuDetectEvent event.

And the javadoc is inconsistent:

Event: widget-relative
MenuDetectEvent:  display-relative

but all the conversion does is
```
public MenuDetectEvent(Event e) {
	super(e);
	this.x = e.x;
	this.y = e.y;
	this.doit = e.doit;
	this.detail = e.detail;
}
```

**Pull request merged**